cmd/kubeadm/app/apis/kubeadm/validation/validation.go

```diff
@@ -52,57 +54,283 @@ func ValidateMasterConfiguration(c *kubeadm.MasterConfiguration) field.ErrorList
+				),
+			)
+		}
+	}
+	return allErrs
+}
+
+func ValidateTLSBootstrapTokenArg(cfg *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(cfg.TLSBootstrapToken) != 0 {
+		if len(cfg.Token) != 0 || len(cfg.DiscoveryURL) != 0 || len(cfg.DiscoveryFile) != 0 {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					fldPath,
+					nil,
+					"--tls-bootstrap-token is mutually exclusive with --token, --discovery-file, and --discovery-url",
+				),
+			)
+		}
+		if len(cfg.Masters) < 1 {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					fldPath,
+					nil,
+					"--tls-bootstrap-token must also specificy --masters",
+				),
+			)
+		}
+		if len(cfg.DiscoveryToken) != 0 {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					fldPath,
+					nil,
+					"--tls-bootstrap-token must also specificy --discovery-token",
+				),
+			)
+		}
+	}
+	return allErrs
+}
+
+func ValidateDiscoveryFileArg(cfg *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(cfg.DiscoveryFile) != 0 {
+		if len(cfg.Token) != 0 || len(cfg.DiscoveryToken) != 0 || len(cfg.DiscoveryURL) != 0 {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					fldPath,
+					nil,
+					"--discovery-file is mutually exclusive with --token, --discovery-token, and --discovery-url",
+				),
+			)
+		}
+		if len(cfg.Masters) != 0 {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					fldPath,
+					nil,
+					"--discovery-file must not specificy --masters",
+				),
+			)
+		}
+		if len(cfg.TLSBootstrapToken) != 0 {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					fldPath,
+					nil,
+					"--discovery-file must also specificy --tls-bootstrap-token",
+				),
+			)
+		}
+	}
 	return allErrs
 }
 
-func ValidateHTTPSDiscovery(c *kubeadm.HTTPSDiscovery, fldPath *field.Path) field.ErrorList {
+func ValidateDiscoveryURLArg(cfg *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
+	if len(cfg.DiscoveryURL) != 0 {
+		if len(cfg.Token) != 0 || len(cfg.DiscoveryToken) != 0 || len(cfg.DiscoveryFile) != 0 {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					fldPath,
+					nil,
+					"--discovery-url is mutually exclusive with --token, --discovery-token, and --discovery-file",
+				),
+			)
+		}
+		if len(cfg.Masters) != 0 {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					fldPath,
+					nil,
+					"--discovery-url must not specificy --masters",
+				),
+			)
+		}
+		if len(cfg.TLSBootstrapToken) != 0 {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					fldPath,
+					nil,
+					"--discovery-url must also specificy --tls-bootstrap-token",
+				),
+			)
+		}
+	}
 	return allErrs
 }
 
-func ValidateTokenDiscovery(c *kubeadm.TokenDiscovery, fldPath *field.Path) field.ErrorList {
+func ValidateTLSBootstrapToken(c *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
```

Comment:
`ValidateToken(token string, ...)` and run this for `--token`, `--discovery-token` and `--tls-bootstrap-token` respectively if present